src/server/internalServer.cpp

```diff
@@ -451,28 +453,49 @@ bool InternalServer::start() {
   if (m_verbose.load())
     flags |= MHD_USE_DEBUG;
 
-  struct sockaddr_in sockAddr;
-  memset(&sockAddr, 0, sizeof(sockAddr));
-  sockAddr.sin_family = AF_INET;
-  sockAddr.sin_port = htons(m_port);
-  if (m_addr.empty()) {
-    if (0 != INADDR_ANY) {
-      sockAddr.sin_addr.s_addr = htonl(INADDR_ANY);
-    }
-    m_addr = kiwix::getBestPublicIp();
+
+  struct sockaddr_in sockAddr4={0};
+  struct sockaddr_in6 sockAddr6={0};
+
+
+  if (m_ipv6) {
+    flags|=MHD_USE_DUAL_STACK;
+    sockAddr6.sin6_family = AF_INET6;
+    sockAddr6.sin6_port = htons(m_port);
+    if (m_addr.empty()) {
+      if (0 != INADDR_ANY) {
+        sockAddr6.sin6_addr = in6addr_any;
+      }
+      m_addr = (kiwix::getBestPublicIp(m_ipv6)).addr6;
+    } else {
+      if (inet_pton(AF_INET6, m_addr.c_str(), &(sockAddr6.sin6_addr.s6_addr)) == 0) {
+        std::cerr << "Ip address " << m_addr << "  is not a valid ipv6 address" << std::endl;
+        return false;
+      }
+      }
   } else {
-    if (inet_pton(AF_INET, m_addr.c_str(), &(sockAddr.sin_addr.s_addr)) == 0) {
-      std::cerr << "Ip address " << m_addr << "  is not a valid ip address" << std::endl;
+    sockAddr4.sin_family = AF_INET;
+    sockAddr4.sin_port = htons(m_port);
+    if (m_addr.empty()) {
+      if (0 != INADDR_ANY) {
+        sockAddr4.sin_addr.s_addr = htonl(INADDR_ANY);
+      }
+      m_addr = (kiwix::getBestPublicIp(m_ipv6)).addr;
+    } else{
```

Comment:
I would move this piece of code along with its counterpart in the other branch of the enclosing `if` statement outside (before) that statement. Of course it will have to initialize both `sockAddr4` and `sockAddr6` but as a result overall nesting will be reduced.